Add nested rowtype reference support

joy-yao opened the pull request.

cberner commented:
"dereference" should we spelled as one word throughout the code

cberner commented:
back to you for comments

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java:
```diff
@@ -103,11 +107,17 @@ public AggregationAnalyzer(List<FieldOrExpression> groupByExpressions, Metadata
+                name = ((QualifiedNameReference) expression).getName();
+            }
+            else {
+                name = DereferenceExpression.getQualifiedName((DereferenceExpression) expression);
```
Use `checkType` instead of this cast, so that we'll get a nice error message, if there's some how a different type of Expression here

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java:
```diff
@@ -334,11 +351,21 @@ public Boolean visitWindowFrame(WindowFrame node, Void context)
-            Preconditions.checkState(!fields.isEmpty(), "No fields for name '%s'", name);
-            Preconditions.checkState(fields.size() <= 1, "Found more than one field for name '%s': %s", name, fields);
+        @Override
+        protected Boolean visitDereferenceExpression(DereferenceExpression node, Void context)
```
This doesn't handle dereferences which are row field references correctly. For example, try this test case:

```
 assertQuery("SELECT a.col1[1].col0, SUM(a.col0), SUM(a.col1[1].col1), SUM(a.col1[2].col0), SUM(a.col2.col1) FROM " +
                        "(VALUES " +
                        "(ROW(test_row(1.0, ARRAY[test_row(31, 4.5), test_row(12, 4.2)], test_row(3, 4.0)))), " +
                        "(ROW(test_row(3.1, ARRAY[test_row(41, 3.1), test_row(32, 4.2)], test_row(6, 6.0)))), " +
                        "(ROW(test_row(2.2, ARRAY[test_row(31, 4.2), test_row(22, 4.2)], test_row(5, 4.0))))) t(a) " +
                        "GROUP BY a.col1[1]",
                "SELECT * FROM VALUES (31, 3.2, 8.7, 34, 8.0), (41, 3.1, 3.1, 32, 6.0)");
```

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -152,15 +153,24 @@ public static Object evaluateConstantExpression(Expression expression, Type expe
+            {
+                if (columnReferences != null && columnReferences.contains(node)) {
+                    throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain column references");
+                }
```
You need to visit the base in the case that it's a row field reference. Otherwise `a.col0` wouldn't throw an exception if `a` is a row type column

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java:
```diff
@@ -134,6 +144,13 @@ public void analyze(Expression expression)
     private class Visitor
             extends AstVisitor<Boolean, Void>
     {
+        private final Set<Expression> columnReferences;
```
Since this is a non-static inner class, you don't need this field. You can just access `columnReferences` in the parent class

cberner commented:
There are a couple edge cases you missed. Had a couple other minor comments too, so just put the updates in a separate commit and I'll review again :)

cberner commented:
@joy-yao this PR fixes https://github.com/facebook/presto/issues/3371, right?

joy-yao commented:
Thanks Chris, updated with your review comments. Also fixes two issues -- one is with row field in JOIN ON clause, and the other is the one you pointed out SELECT SUM(a.col1[1].col1) ... GROUP BY a.col1[1]. Added test cases for both cases.

joy-yao commented:
@cberner , yes, this fixes  #3371.

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -284,61 +285,68 @@ protected Type visitQualifiedNameReference(QualifiedNameReference node, Analysis
+
+            Type baseType = process(node.getBase(), context);
+            if (!(baseType instanceof RowType)) {
+                throw new SemanticException(SemanticErrorCode.TYPE_MISMATCH, node.getBase(), "Expression %s is not RowType", node.getBase());
```
I think the user will see this message, so change it to say "Expression %s is not of type ROW".

cberner commented:
Had a minor comment. Looks good.
Wait for the release though, and remember to squash the commits.

joy-yao commented:
Thanks a lot @cberner ! Addressed your comment, squashed and will merge after the release : )
